### ISSUE TITLE
Error code is not verbose enough to find reason for auth failure

### ISSUE THREAD

**Issue by tipa** [AzureAD/microsoft-authentication-library-for-dotnet#673]
**Which Version of MSAL are you using ?**
MSAL 2.3.1-preview

**Which platform has the issue?**
Xamarin.Android

**What authentication flow has the issue?**
Xamarin.Android

**Repro**

```csharp
var clientApplication = new PublicClientApplication(ClientId) { RedirectUri = "msal...xyz...://auth" };
authResult = await clientApplication.AcquireTokenSilentAsync(Scopes, (await clientApplication.GetAccountsAsync()).FirstOrDefault());
```

**Expected behavior**
User can finish the auth process either successfully or a meaningful error is returned

**Actual behavior**
For an increasing number of my users, the auth process fails with the error code "invalid_grant". How can this issue be solved from either an app- oder user-perspective? Ideally, I would like to see the library to give a more meaningful/verbose error description.
This may be related to https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/559

**Additional context/ Logs / Screenshots**
```
(False) MSAL 2.3.1.0 MSAL.Xamarin.Android 22 [11/04/2018 21:30:11] MSAL MSAL.Xamarin.Android with assembly version '2.3.1.0', file version '2.3.1.0' and informational version '2.3.1-preview' is running... 
(False) MSAL 2.3.1.0 MSAL.Xamarin.Android 22 [11/04/2018 21:30:12 - 1068f1c4-cf72-4d1f-8a63-51f40748a7c8] === Request Data === Authority Provided? - True Scopes - Files.ReadWrite.AppFolder Validate Authority? - True LoginHint provided? - False User provided? - False Extra Query Params Keys (space separated) - Slice Parameters Keys(space separated) - 
(False) MSAL 2.3.1.0 MSAL.Xamarin.Android 22 [11/04/2018 21:30:12 - 1068f1c4-cf72-4d1f-8a63-51f40748a7c8] Additional scopes - ;UIBehavior - select_account 
(False) MSAL 2.3.1.0 MSAL.Xamarin.Android 22 [11/04/2018 21:30:12 - 1068f1c4-cf72-4d1f-8a63-51f40748a7c8] === Token Acquisition (InteractiveRequest) started: Cache Provided: True Authority Host: login.microsoftonline.com 
(False) MSAL 2.3.1.0 MSAL.Xamarin.Android 22 [11/04/2018 21:30:12 - 1068f1c4-cf72-4d1f-8a63-51f40748a7c8] Resolving authority endpoints... Already resolved? - False 
(False) MSAL 2.3.1.0 MSAL.Xamarin.Android 22 [11/04/2018 21:30:12 - 1068f1c4-cf72-4d1f-8a63-51f40748a7c8] Is Authority tenantless? - True 
(False) MSAL 2.3.1.0 MSAL.Xamarin.Android 22 [11/04/2018 21:30:14] MSAL MSAL.Xamarin.Android with assembly version '2.3.1.0', file version '2.3.1.0' and informational version '2.3.1-preview' is running... 
(False) MSAL 2.3.1.0 MSAL.Xamarin.Android 22 [11/04/2018 21:31:03] Received Activity Result(2003) 
(False) MSAL 2.3.1.0 MSAL.Xamarin.Android 22 [11/04/2018 21:31:04] MSAL MSAL.Xamarin.Android with assembly version '2.3.1.0', file version '2.3.1.0' and informational version '2.3.1-preview' is running... 
(False) MSAL 2.3.1.0 MSAL.Xamarin.Android 22 [11/04/2018 21:32:44 - 1068f1c4-cf72-4d1f-8a63-51f40748a7c8] Exception type: System.Threading.Tasks.TaskCanceledException at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00026] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in :0 at System.Net.Http.HttpClientHandler+d__64.MoveNext () [0x004f5] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0001a] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at System.Net.Http.HttpClient+d__49.MoveNext () [0x000ca] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0001a] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at Microsoft.Identity.Core.Http.HttpManager+d__11.MoveNext () [0x000cf] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0001a] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at Microsoft.Identity.Core.Http.HttpManager+d__10.MoveNext () [0x00136] in :0 
(False) MSAL 2.3.1.0 MSAL.Xamarin.Android 22 [11/04/2018 21:32:44 - 1068f1c4-cf72-4d1f-8a63-51f40748a7c8] Retrying one more time.. 
(False) MSAL 2.3.1.0 MSAL.Xamarin.Android 22 [11/04/2018 21:32:46 - 1068f1c4-cf72-4d1f-8a63-51f40748a7c8] Response status code does not indicate success: 400 (BadRequest). 
(False) MSAL 2.3.1.0 MSAL.Xamarin.Android 22 [11/04/2018 21:32:46 - 1068f1c4-cf72-4d1f-8a63-51f40748a7c8] Exception type: Microsoft.Identity.Client.MsalUiRequiredException , ErrorCode: invalid_grant , StatusCode: 400 at Microsoft.Identity.Core.OAuth2.OAuth2Client.CreateErrorResponse (Microsoft.Identity.Core.Http.HttpResponse response, Microsoft.Identity.Core.RequestContext requestContext) [0x0003d] in :0 at Microsoft.Identity.Core.OAuth2.OAuth2Client.CreateResponse[T] (Microsoft.Identity.Core.Http.HttpResponse response, Microsoft.Identity.Core.RequestContext requestContext, System.Boolean addCorrelationId) [0x0000d] in :0 at Microsoft.Identity.Core.OAuth2.OAuth2Client+d__9`1[T].MoveNext () [0x00296] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at Microsoft.Identity.Core.OAuth2.OAuth2Client+d__8.MoveNext () [0x0007d] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at Microsoft.Identity.Client.Internal.Requests.RequestBase+d__24.MoveNext () [0x000b2] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at Microsoft.Identity.Client.Internal.Requests.RequestBase+d__23.MoveNext () [0x00159] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at Microsoft.Identity.Client.Internal.Requests.InteractiveRequest+d__9.MoveNext () [0x00160] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at Microsoft.Identity.Client.Internal.Requests.RequestBase+d__18.MoveNext () [0x000df] in :0 
(False) MSAL 2.3.1.0 MSAL.Xamarin.Android 22 [11/04/2018 21:32:46] MSAL MSAL.Xamarin.Android with assembly version '2.3.1.0', file version '2.3.1.0' and informational version '2.3.1-preview' is running... Microsoft.Identity.Client.MsalUiRequiredException: AADSTS70000: The provided value for the 'code' parameter is not valid. The code has expired. Trace ID: bc34191b-7b99-44fa-8473-489d16af5100 Correlation ID: 1068f1c4-cf72-4d1f-8a63-51f40748a7c8 Timestamp: 2018-11-04 21:32:46Z at Microsoft.Identity.Core.OAuth2.OAuth2Client.CreateErrorResponse (Microsoft.Identity.Core.Http.HttpResponse response, Microsoft.Identity.Core.RequestContext requestContext) [0x0003d] in :0 at Microsoft.Identity.Core.OAuth2.OAuth2Client.CreateResponse[T] (Microsoft.Identity.Core.Http.HttpResponse response, Microsoft.Identity.Core.RequestContext requestContext, System.Boolean addCorrelationId) [0x0000d] in :0 at Microsoft.Identity.Core.OAuth2.OAuth2Client+d__9`1[T].MoveNext () [0x00296] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at Microsoft.Identity.Core.OAuth2.OAuth2Client+d__8.MoveNext () [0x0007d] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at Microsoft.Identity.Client.Internal.Requests.RequestBase+d__24.MoveNext () [0x000b2] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at Microsoft.Identity.Client.Internal.Requests.RequestBase+d__23.MoveNext () [0x00159] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at Microsoft.Identity.Client.Internal.Requests.InteractiveRequest+d__9.MoveNext () [0x00160] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at Microsoft.Identity.Client.Internal.Requests.RequestBase+d__18.MoveNext () [0x00150] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at Microsoft.Identity.Client.PublicClientApplication+d__12.MoveNext () [0x00106] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1+ConfiguredTaskAwaiter[TResult].GetResult () [0x00000] in :0 at Microsoft.Identity.Client.PublicClientApplication+d__4.MoveNext () [0x0009b] in :0 --- End of stack trace from previous location where exception was thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in :0 at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in :0 at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in :0 at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in :0 at Diarium.OneDriveHelper+d__0.MoveNext () [0x00349] in :0 ErrorCode: invalid_grant StatusCode: 400 Claims: 
```

**Comment by henrik-me** [comment]
We are working with the service side bring more detailed error codes to the client library. We haven't settled on everything yet and thus we can't speak to the timeline. However this has high priority for us.

**Comment by henrik-me** [comment]
Currently keeping this open, however this is as far as I can see the same ask as what you linked to. Though there might be some thing additional we can/should do for Xamarin, thus keeping it open for now.

**Comment by jmprieur** [comment]
Closing as duplicate of #173  #559 (fixed in #841)
